Add verified personal information create/update GraphQL query

akikoskinen opened the pull request:
This PR adds a concept of "verified personal information" into Profile. The GraphQL API is fully present. There's a single GraphQL mutation that handles both creation and updates.

This PR **doesn't contain** input validation, permission handling, nor returning of the verified personal information via any GraphQL query.

HP-306

tituomin commented on profiles/schema.py:
```diff
@@ -688,6 +693,102 @@ def mutate_and_get_payload(cls, root, info, **input):
+                address_type["name"], None
+            )
+
+        (user, created) = User.objects.get_or_create(uuid=user_id_input)
```
I don't think the parentheses are needed.

akikoskinen replied:
Removed

tituomin reviewed (changes requested):
Security review checklist added to wiki.

akikoskinen commented on profiles/models.py:
```diff
@@ -237,6 +237,58 @@ def import_customer_data(cls, data):
+    )
+
+    @classmethod
+    def post_delete(cls, instance, using, **kwargs):
```
Check the commit message

tituomin replied:
I read it now, but it doesn't explain why on_delete=models.CASCADE doesn't work.

tituomin replied:
Oh, sorry, the direction would be wrong there.

akikoskinen replied:
> Oh, sorry, the direction would be wrong there.

Yep, that's right. But look at the data model closely. Is this the right way to do it? Usually I see `OneToOne` fields pointing "the other way". So in here the `OneToOne` field would be in `EncryptedAddress`, pointing to `VerifiedPersonalInformation`. But off course that doesn't work right here, because that wouldn't tell if a given address is the permanent or the temporary address. But is there some other solution for this kind of data?

tituomin replied:
The same thing crossed my mind. But I didn't mentioned it because I realized you wanted to reuse the same class for both fields.

The way to do it i guess would be to make EncryptedAddress a base class and add 2 subclasses with different one-to-one fields and different related_names

tituomin replied:
Your original solution would be much cleaner if not for the extra deletion code.

akikoskinen replied:
Hmm, that sounds like it could work 🤔

tituomin replied:
Yeah, kinda sad it would obscure the main model by sprinkling its field definitions elsewhere. OTOH maybe that's the Django Way. :)

akikoskinen replied:
I updated this PR with the way discussed in this thread. I think it is the Django way as Django now takes care of some things that required extra code in the previous version. There is one more model and database table with this approach, but perhaps that's not too bad.

akikoskinen commented on open_city_profile/tests/snapshots/snap_test_graphql_api_schema.py:
```diff
@@ -727,6 +741,24 @@
+input VerifiedPersonalInformationAddressInput {
+  streetName: String
+  postalCode: String
+  city: String
```
I used the same terminology that is already used in [Address](https://github.com/City-of-Helsinki/open-city-profile/blob/develop/profiles/models.py#L310). I agree it's not the best term, but I used it for consistency.

Then again, for the national identification number, I didn't use `ssn` which is already used in [SensitiveData](https://github.com/City-of-Helsinki/open-city-profile/blob/develop/profiles/models.py#L250). So I guess I wasn't that consistent after all 🤔 

I think I'm fine with not being consistent with poorly chosen names.

akikoskinen replied:
Renamed `city` to `post office`.

Also renamed `street name` to `street address` because that field, in addition to street name, may also contain e.g. house number, apartment number and such. All of those together is called `street address`.

tkutinla commented:
According to https://helsinkisolutionoffice.atlassian.net/browse/HP-319, Address may contain country code ("Vakinainen ulkomainen lähiosoite, valtiokoodi").  The Suomi.fi documentation, https://palveluhallinta.suomi.fi/fi/tuki/artikkelit/590ad07b14bbb10001966f50, talks about "Vakinaisen ulkomaisen osoitteen valtiokoodi."

Should the VerifiedPersonalInformationAddressInput  include "countryCode" -field?  Now it does not, it only has fields: streetName, postalCode, city

akikoskinen commented:
> According to https://helsinkisolutionoffice.atlassian.net/browse/HP-319, Address may contain country code ("Vakinainen ulkomainen lähiosoite, valtiokoodi"). The Suomi.fi documentation, https://palveluhallinta.suomi.fi/fi/tuki/artikkelit/590ad07b14bbb10001966f50, talks about "Vakinaisen ulkomaisen osoitteen valtiokoodi."
> 
> Should the VerifiedPersonalInformationAddressInput include "countryCode" -field? Now it does not, it only has fields: streetName, postalCode, city

I have been following [this specification](https://helsinkisolutionoffice.atlassian.net/wiki/spaces/KAN/pages/460521553/Profiilitietojen+hallinta). I doesn't contain any address abroad.

tkutinla commented:
> I have been following [this specification](https://helsinkisolutionoffice.atlassian.net/wiki/spaces/KAN/pages/460521553/Profiilitietojen+hallinta). It doesn't contain any address abroad.

According to Valtteri Kosonen, the latest known details are in https://helsinkisolutionoffice.atlassian.net/browse/HP-319. not in Confluence.

tkutinla commented:
The fields in verifedPersonalInformation like firtsName, lastName. etc.  are of type _`EnryptedCharField`_.

Are these fields searchable.  Is it possible to implement GraphQL query  `return profiles where verifiedPersonalInformation.firstName equals 'Timo'``.  Or "... `where lastName starts with 'T'`

We have to verify from product owner if these fields should be searchable or not.

akikoskinen commented:
> The fields in verifedPersonalInformation like firtsName, lastName. etc. are of type _`EnryptedCharField`_.
> 
> Are these fields searchable. Is it possible to implement GraphQL query `` return profiles where verifiedPersonalInformation.firstName equals 'Timo'``. Or "... ``where lastName starts with 'T'`

Not in the same way at least as unencrypted fields. The library providing these fields is called [django-searchable-encrypted-fields](https://pypi.org/project/django-searchable-encrypted-fields/). As the name suggests,  there's some searchability available. It's possible to search using these fields by **exact match only**. So no "starts with" queries.

The searchability also needs to be enabled separately for each encrypted field that needs it. In this PR none of the encrypted fields are searchable.

akikoskinen commented:
I returned another time to this PR. I changed the GraphQL mutation API's input data type a bit. Previously it looked something like this:

```graphql
mutation {
    createOrUpdateProfileWithVerifiedPersonalInformation(
        input: {
            userId: <UUID>,
            verifiedPersonalInformation: {
                firstName: "John",
            },
        }
    )
}
```

I added a `profile` object as a parent to the `verifiedPersonalInformation` object so the API is now more like this:

```graphql
mutation {
    createOrUpdateProfileWithVerifiedPersonalInformation(
        input: {
            userId: <UUID>,
            profile: {
                verifiedPersonalInformation: {
                    firstName: "John",
                },
            }
        }
    )
}
```

This change makes this new API similar to the other profile mutation APIs.
